driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java

```diff
@@ -96,108 +113,137 @@ public void connect() {
+                        || (currentServerDescription.getException() instanceof MongoSocketException
+                        && previousServerDescription.getType() != UNKNOWN)) {
+                    continue;
                 }
+                waitForNext();
             }
         }
 
         private ServerDescription getConnectingServerDescription(final Throwable exception) {
             return ServerDescription.builder().type(UNKNOWN).state(CONNECTING).address(serverId.getAddress()).exception(exception).build();
         }
 
-        private ServerDescription lookupServerDescription(final InternalConnection connection) {
-            if (LOGGER.isDebugEnabled()) {
-                LOGGER.debug(format("Checking status of %s", serverId.getAddress()));
-            }
-            serverMonitorListener.serverHearbeatStarted(new ServerHeartbeatStartedEvent(connection.getDescription().getConnectionId()));
-
-            long start = System.nanoTime();
+        private ServerDescription lookupServerDescription(final ServerDescription currentServerDescription) {
             try {
-                BsonDocument isMasterResult =
-                        executeCommand("admin", new BsonDocument("ismaster", new BsonInt32(1)), clusterClock, connection);
-                long elapsedTimeNanos = System.nanoTime() - start;
-                averageRoundTripTime.addSample(elapsedTimeNanos);
-
-                serverMonitorListener.serverHeartbeatSucceeded(
-                        new ServerHeartbeatSucceededEvent(connection.getDescription().getConnectionId(), isMasterResult, elapsedTimeNanos));
-
-                return createServerDescription(serverId.getAddress(), isMasterResult, averageRoundTripTime.getAverage());
-            } catch (RuntimeException e) {
-                serverMonitorListener.serverHeartbeatFailed(
-                        new ServerHeartbeatFailedEvent(connection.getDescription().getConnectionId(), System.nanoTime() - start, e));
-                throw e;
+                if (connection == null || connection.isClosed()) {
+                    connection = internalConnectionFactory.create(serverId);
+                    connection.open();
+                    averageRoundTripTime.addSample(connection.getInitialServerDescription().getRoundTripTimeNanos());
+                    return connection.getInitialServerDescription();
+                }
+
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(format("Checking status of %s", serverId.getAddress()));
+                }
+                serverMonitorListener.serverHearbeatStarted(new ServerHeartbeatStartedEvent(connection.getDescription().getConnectionId()));
+
+                long start = System.nanoTime();
+                try {
+                    SessionContext sessionContext = new ClusterClockAdvancingSessionContext(NoOpSessionContext.INSTANCE, clusterClock);
+                    if (!connection.hasMoreToCome()) {
```

Comment:
This part looks a bit different that the pseudocode in the spec, but might be a bit confusing.  The key point is that we're separating `send` and `receive` regardless of whether we're doing the streaming protocol.   The first time we get here, `moreToCome` will be false, so we send an isMaster, possibly using streaming protocol.  Then we immediately receive the first response.  The second time we get here (from the main loop above), `moreToCome` will be true if we're streaming, and we won't send, we'll just receive.  But if we aren't streaming, `moreToCome` will be false, and we'll both send and receive.